### PR TITLE
Move Mode!

### PR DESCRIPTION
**Pull request by DawnTheWitch**

### CODE REVIEW

File: src/MoveSingleMode.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            const tempCut: CutNode = new CutNode(
+                new Ellipse(
+                    new Point(
+                        currentNode.ellipse.center.x + moveDifference.x - offset.x,
+                        currentNode.ellipse.center.y + moveDifference.y - offset.y
+                    ),
+                    currentNode.ellipse.radiusX,
+                    currentNode.ellipse.radiusY
+                )
+            );
```
Comment (James-Oswald) [inline src/MoveSingleMode.ts:137]: Helper function helper function helper function

File: src/MoveSingleMode.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        } else if (currentNode instanceof AtomNode) {
+            const tempAtom: AtomNode = new AtomNode(
+                currentNode.identifier,
+                new Point(
+                    currentNode.origin.x + moveDifference.x - offset.x,
+                    currentNode.origin.y + moveDifference.y - offset.y
+                ),
+                currentNode.width,
+                currentNode.height
+            );
```
Comment (James-Oswald) [inline src/MoveSingleMode.ts:154]: Helper function helper function helper function

File: src/MoveSingleMode.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            const tempCut: CutNode = new CutNode(
+                new Ellipse(
+                    new Point(
+                        currentNode.ellipse.center.x + moveDifference.x - offset.x,
+                        currentNode.ellipse.center.y + moveDifference.y - offset.y
+                    ),
+                    currentNode.ellipse.radiusX,
+                    currentNode.ellipse.radiusY
+                )
+            );
```
Comment (James-Oswald) [inline src/MoveSingleMode.ts:80]: Helper function helper function helper function

File: src/MoveMultiMode.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        } else if (currentNode instanceof AtomNode) {
+            const tempAtom: AtomNode = new AtomNode(
+                currentNode.identifier,
+                new Point(
+                    currentNode.origin.x + moveDifference.x - offset.x,
+                    currentNode.origin.y + moveDifference.y - offset.y
+                ),
+                currentNode.width,
+                currentNode.height
+            );
```
[the remark this answers is not in the document]
Comment (James-Oswald) [reply src/MoveMultiMode.ts:117]: Helper function helper function helper function

File: src/MoveMultiMode.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
```
[the remark this answers is not in the document]
Comment (James-Oswald) [reply src/MoveMultiMode.ts:77]: Helper function helper function helper function

File: src/MoveMultiMode.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+ * @param event The mouse down event while in moveMulti mode.
+ */
+export function moveMultiMouseDown(event: MouseEvent) {
+    startingPoint = new Point(event.x - offset.x, event.y - offset.y);
+    if (tree.getLowestNode(startingPoint) !== tree.sheet) {
+        currentNode = tree.getLowestNode(startingPoint);
+        const currentParent = tree.getLowestParent(startingPoint);
+        currentParent.remove(startingPoint);
+        legalNode = true;
+    }
```
[the remark this answers is not in the document]
Comment (James-Oswald) [reply src/MoveMultiMode.ts:44]: You should also ensure that currentNode has a non-undefined default value here... `null` would be a good choice.

File: src/index.ts [hunk 7]
```diff
@@ -73,31 +120,84 @@ modeButtons.forEach(button => {
...
+}
+
+/**
+ * Sets the current mode to deleteMultiMode. Hides non delete tools.
+ */
+function deleteMultiMode() {
+    modeState = Mode.deleteMultiMode;
     cutTools.style.display = "none";
     atomTools.style.display = "none";
 }
```
Comment (James-Oswald) [inline src/index.ts:203]: Too much code reuse. Combine all of these into a single function `setMode` that takes a mode Enum for the mode. Export the Mode Enum, and from the HTML button onClicks call `setMode(Mode.somethingMode)` for each mode button.

File: src/ModeUtils.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        const tempCut: CutNode = new CutNode(
+            new Ellipse(
+                new Point(
+                    incomingNode.ellipse.center.x + change.x - offset.x,
+                    incomingNode.ellipse.center.y + change.y - offset.y
+                ),
+                incomingNode.ellipse.radiusX,
+                incomingNode.ellipse.radiusY
+            )
+        );
```
[the remark this answers is not in the document]
Comment (DawnTheWitch) [reply src/ModeUtils.ts:88]: Are we sure it shouldn't be exported? I use this in the mode files themselves as well.

[the remark this answers is not in the document]
Comment (James-Oswald) [reply src/ModeUtils.ts:88]: It should be exported, see later comments

[the remark this answers is not in the document]
Comment (DawnTheWitch) [reply src/ModeUtils.ts:88]: I did see those I just wanted to make sure.

File: src/AEG/CutNode.ts [hunk 9]
```diff
@@ -122,6 +122,63 @@ export class CutNode {
...
+     * @returns The parent of the lowest level node.
+     */
+    public getLowestParent(incomingPoint: Point): CutNode {
+        for (let i = 0; i < this.internalChildren.length; i++) {
+            if (this.internalChildren[i].containsPoint(incomingPoint)) {
+                //If there are no children this is the lowest node.
+                if (
+                    this.internalChildren[i] instanceof AtomNode ||
+                    (this.internalChildren[i] instanceof CutNode &&
+                        (this.internalChildren[i] as CutNode).children.length === 0)
```
[the remark this answers is not in the document]
Comment (DawnTheWitch) [reply src/AEG/CutNode.ts:162]: Lint error is why it's like this, I think it looks ugly as sin
image
image

**Pull request merged**